### PR TITLE
[BugFix] pp cannot run successfully under NixlConnector

### PR DESCRIPTION
**Pull request by R2-Y**
[BugFix] pp cannot run successfully under NixlConnector 

## Purpose
I previously raised issue #22430 regarding NixlConnector failing with Pipeline Parallelism (PP) + Prefill-Decode (PD) configurations, but it hasn't received attention from the community yet. Since I need PP functionality for my project and my benchmarks show that NixlConnector delivers significantly better performance than LMCacheConnector, I'm motivated to contribute a fix.

My testing on Qwen 2.5 7B demonstrates NixlConnector's clear advantage:

Configuration | TTFT (ms) | Improvement
-- | -- | --
PP1 + TP2 + PD + LMCache | 13,900.61 | Baseline
PP1 + TP2 + PD + Nixl | 10,333.13 | 25.7% faster

The error occurs during KV cache initialization when ```nixl_wrapper.register_memory()``` is called in a Pipeline Parallelism setup. After investigating the codebase, I discovered that the current implementation only considers Tensor Parallelism (TP) when initializing KV caches. 
image
image

When registering cache data with NIXL, different PP stages that share the same TP rank are assigned to the same device ID. This causes device conflicts during PP memory registration, leading to the ```NIXL_ERR_BACKEND``` failure. Therefore, I modified the device ID assignment from using TP rank to automatically detecting the current device using ```torch.cuda.current_device()```.

## Test Plan
Tested by running the model under PP + NixlConnector for both PD disaggregation and PD aggregation scenarios.

## Test Result

The error no longer occurs.

## (Optional) Documentation Update

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

**Comment by R2-Y** [comment]
@WoosukKwon @robertgshaw2-redhat @lk-chen Hi could you help to review?

**Review by lk-chen — APPROVED** [review]

**Comment by lk-chen** [comment]
You may want to add some basic tests in `test_nixl_connector.py`, `TestNixlHandshake` has some good examples.

and cc @NickLucche

**Comment by njhill** [comment]
Thanks @R2-Y, we should make sure that this is compatible with:
- Heterogeneous TP
- TPU

I have a feeling some changes might be required for those cc @juncgu

**Comment by R2-Y** [comment]
> Thanks @R2-Y, we should make sure that this is compatible with:
> 
> * Heterogeneous TP
> * TPU
> 
> I have a feeling some changes might be required for those cc @juncgu

Hi Nick, Thanks for your review.  I tested the heterogeneous TP + PP model with PD disaggregation (1P TP2PP2; 1D TP1PP2), and it worked well, with correct inference results. However, I'm still doing more testing to see if it works well with other configurations.  For the TPU, I will add a method to get the device ID later.

File: vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py [hunk 1]
```diff
@@ -238,7 +238,8 @@ def __init__(self, vllm_config: VllmConfig, engine_id: str):
         self.side_channel_port = (
             envs.VLLM_NIXL_SIDE_CHANNEL_PORT +
             vllm_config.parallel_config.data_parallel_rank *
-            vllm_config.parallel_config.tensor_parallel_size)
+            vllm_config.parallel_config.tensor_parallel_size *
+            vllm_config.parallel_config.pipeline_parallel_size)
```
Comment (NickLucche) [inline vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py:244]: just a note: this is going to be something we will happily manage better with @wseaton http based PR. But it will conflict with the dp/tp endpoint format.

File: vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py [hunk 2]
```diff
@@ -439,11 +440,15 @@ def __init__(self, vllm_config: VllmConfig, engine_id: str):
         # NIXL handshake port.
         # NOTE(rob): Within a DP group, each DP rank gets its own
         # base port (which is sent in the KVTransferParams).
-        # Each TP rank listens/queries on the base_port + tp_rank.
+        # Each TP rank listens/queries on the base_port +
+        # pp_rank * tp_size + tp_rank.
+        self.pp_rank = get_pipeline_model_parallel_rank()
```
Comment (NickLucche) [inline vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py]: can this be done with `world_rank`?
Doesn't world-level parallelism coordinate take PP into accout?

File: vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py [hunk 3]
```diff
@@ -545,6 +550,7 @@ def __init__(self, vllm_config: VllmConfig, engine_id: str):
         logger.debug("Detected kv cache layout %s", self.kv_cache_layout)
 
         self._tp_size: dict[EngineId, int] = {self.engine_id: self.world_size}
+        self.device_id = torch.cuda.current_device()
```
Comment (NickLucche) [inline vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py]: I think we definitely need to use some abstraction here to get the device

Comment (NickLucche) [reply vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py]: eg it will break tpu

**Review by NickLucche — CHANGES_REQUESTED** [review]
Hey @R2-Y thanks a lot for the contribution!
I think the logic of your proposals is right in the sense that we need to add that extra PP dimension to our indexing.
 
Off the top of my head though this won't work with heteroTP due to (at least) the fact that we're not accounting for pp when handshaking

```
        tp_ratio = self._tp_size[self.engine_id] // remote_tp_size
        p_remote_rank = self.tp_rank // tp_ratio
```
But there might also be other constraints to sharding the kv between tp ranks and pp stages. Will need to clear that out in some design doc likely.

To that extent, I would start by adding PP dim in `run_accuracy_test`.

Also, I believe we need to have some sort of abstraction when computing the indexing with dp*tp*pp.
Can the world_parallel coordinates help out here in getting unique indices without keeping all the logic in the connector?

We can even look into adding PP+heteroTP support in a separate PR, as long as this one does not break the non-PP case.

**Comment by IEI-mjx** [comment]
> > Thanks @R2-Y, we should make sure that this is compatible with:
> > 
> > * Heterogeneous TP
> > * TPU
> > 
> > I have a feeling some changes might be required for those cc @juncgu
> 
> Hi Nick, Thanks for your review. I tested the heterogeneous TP + PP model with PD disaggregation (1P TP2PP2; 1D TP1PP2), and it worked well, with correct inference results. However, I'm still doing more testing to see if it works well with other configurations. For the TPU, I will add a method to get the device ID later.

Hi, @R2-Y Nice work!!
Can you share your PD disaggregation workflow?
Or can this be used in dynamo-vllm_v1 which also use nixlconnector?

**Comment by R2-Y** [comment]
I found some issues about handshake between workers among different PP stage using Ray+nixlconnector, the following stages failed to initialize handshake. In my previous tests,  the inference result correct is because D instance recompute the prefill part although P instance calculated.  Hence, It did not really succeed in performing the PD disaggregation. I will add one more commit to fix this issue later.

**Comment by kahakuka** [comment]
@R2-Y Hello, Does this PP support multiple machines?

**Comment by njhill** [comment]
See also https://github.com/vllm-project/vllm/pull/19591

**Review by NickLucche — CHANGES_REQUESTED** [review]
Should we turn this into an RFC?  
I feel this could use a bit more thinking to make sure we have a design that doesn't prevent a clean heteroTP implementation @njhill

File: vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py (hunk 2 above)
Comment (R2-Y) [reply vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py]: I changed it to global rank within the DP group

File: vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py (hunk 3 above)
Comment (R2-Y) [reply vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py]: use _get_current_device_id instead
```python   
def _get_current_device_id(self) -> int:
        """Get the current device ID in a platform-agnostic way."""
        from vllm.platforms import current_platform

        if current_platform.is_cuda_alike():
            return torch.cuda.current_device()
        elif current_platform.is_tpu():
            return get_tensor_model_parallel_rank()
        elif current_platform.is_xpu():
            try:
                import intel_extension_for_pytorch as ipex
                return ipex.xpu.current_device()
            except ImportError:
                return get_tensor_model_parallel_rank()
        else:
            # For CPU and other platforms
            return get_tensor_model_parallel_rank()
```

**Comment by chaunceyjiang** [comment]
You can take a look at my PR [https://github.com/vllm-project/vllm/pull/19591](https://github.com/vllm-project/vllm/pull/19591). Its design supports PP by adjusting the way metadata is exchanged.

@njhill @R2-Y @NickLucche

**Comment by chaunceyjiang** [comment]
> Should we turn this into an RFC?

+1

**Comment by R2-Y** [comment]
Hi @chaunceyjiang My current modification also supports heterogeneous TP. I noticed that some scheduling logic of PP conflicts with the previous nixlconnector design, so I added some functions for pulling kvcache in the model runner to avoid the problem of ray dag node disorder when kvcache is not ready under PP. So based on my understanding, I suppose this issue still exists in your PR.  Besides, my initial idea was to just modify the device ID in the registry memory so that the device IDs of different PP ranks would not conflict due to the use of same TP ranks, I noticed that you still used tp rank as device id in your modification. If some PP stage are on the same node, you will still encounter problems related to registerMem.

**Comment by chaunceyjiang** [comment]
> I noticed that some scheduling logic of PP conflicts with the previous nixlconnector design, so I added some functions for pulling kvcache in the model runner to avoid the problem of ray dag node disorder when kvcache is not ready under PP

Indeed, I seem to have encountered this issue as well. It appears that this https://github.com/vllm-project/vllm/pull/23770 should fix the problem (not sure).

File: vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py [hunk 4]
```diff
@@ -85,6 +85,7 @@ class ReqMeta:
     remote_port: int
     remote_engine_id: str
     tp_size: int
+    pp_size: int
```
Comment (chaunceyjiang) [inline vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py:89]: There is a scenario with 8 nodes. Four nodes form a P group, and the other four form a D group.  
That is:  
P nodes: PP = 4  
D nodes: PP = 4  

I have a question—for example, how does D node3 exchange nixl metadata with the P nodes?

**Comment by R2-Y** [comment]
> > I noticed that some scheduling logic of PP conflicts with the previous nixlconnector design, so I added some functions for pulling kvcache in the model runner to avoid the problem of ray dag node disorder when kvcache is not ready under PP
> 
> Indeed, I seem to have encountered this issue as well. It appears that this #23770 should fix the problem (not sure).

Yes, I think this can help solve the problem of remote kv cache loading failure when PP>1 in a single node using multiprocessor executor, but I am not sure whether it will work when using Ray executor across nodes. Hi @njhill, can you share whether it can be used with Ray? If so, I will delete my last commit.

File: vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py (hunk 4 above)
Comment (R2-Y) [reply vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py:89]: current version doesn't support multi-node PP, maybe take your PR is a better option to support it.

File: vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py [hunk 5]
```diff
@@ -596,14 +625,16 @@ def _nixl_handshake(
...
+                    f"remote_pp_size {remote_pp_size}, expected_engine_id {expected_engine_id}")
         # Handshake only with the remote TP rank that current local rank will
         # pull from. With homogeneous TP it happens to be the same rank_i.
         tp_ratio = self._tp_size[self.engine_id] // remote_tp_size
-        p_remote_rank = self.tp_rank // tp_ratio
-        path = make_zmq_path("tcp", host, port + p_remote_rank)
-        logger.debug("Querying metadata on path: %s at remote rank %s", path,
-                     p_remote_rank)
+        p_remote_tp_rank = self.tp_rank // tp_ratio
+        p_remote_pp_rank = self.pp_rank # don't support homogeneous PP
```
Comment (ruisearch42) [inline vllm/distributed/kv_transfer/kv_connector/v1/nixl_connector.py]:
```suggestion
        p_remote_pp_rank = self.pp_rank # don't support heterogenous PP
```

File: vllm/distributed/parallel_state.py [hunk 6]
```diff
@@ -306,6 +306,11 @@ def next_rank(self):
         world_size = self.world_size
         return self.ranks[(rank_in_group + 1) % world_size]
 
+    @property
+    def global_rank(self):
```
Comment (ruisearch42) [inline vllm/distributed/parallel_state.py]: Why call it `global_rank()`? Simply `rank()`?

File: vllm/distributed/parallel_state.py [hunk 7]
```diff
@@ -306,6 +306,11 @@ def next_rank(self):
         world_size = self.world_size
         return self.ranks[(rank_in_group + 1) % world_size]
 
+    @property
```
Comment (ruisearch42) [inline vllm/distributed/parallel_state.py:309]: nit: add line above

File: vllm/distributed/parallel_state.py [hunk 8]
```diff
@@ -306,6 +306,11 @@ def next_rank(self):
...
+    def global_rank(self):
+        """Return the current global rank of the process"""
+        return self.rank
```
Comment (ruisearch42) [inline vllm/distributed/parallel_state.py:312]: add line below

File: vllm/distributed/parallel_state.py [hunk 9]
```diff
@@ -1226,6 +1231,18 @@ def get_tensor_model_parallel_rank():
     return get_tp_group().rank_in_group
 
 
+def get_pipeline_model_parallel_rank():
+    """Return my rank for the pipeline model parallel group."""
+    return get_pp_group().rank_in_group
+
+def get_global_rank():
```
Comment (ruisearch42) [inline vllm/distributed/parallel_state.py]: add line above, please fix other places

File: vllm/v1/worker/gpu_model_runner.py [hunk 10]
```diff
@@ -1458,6 +1458,18 @@ def _pool(
             pooler_output=pooler_output,
             kv_connector_output=kv_connector_output,
         )
+    
+    def nixl_pull_kvcache(
```
Comment (ruisearch42) [inline vllm/v1/worker/gpu_model_runner.py]: This is more general, and not mention `nixl`?

Comment (ruisearch42) [inline vllm/v1/worker/gpu_model_runner.py]: Should this be in `KVConnectorModelRunnerMixin`?

File: vllm/v1/worker/gpu_worker.py [hunk 11]
```diff
@@ -350,6 +350,13 @@ def get_model(self) -> nn.Module:
 
     def get_supported_tasks(self) -> tuple[SupportedTask, ...]:
         return self.model_runner.get_supported_tasks()
+    
+    def nixl_pull_kvcache(
```
Comment (ruisearch42) [inline vllm/v1/worker/gpu_worker.py]: ditto, not mention "nixl"?

File: vllm/v1/executor/ray_distributed_executor.py [hunk 12]
```diff
@@ -106,4 +176,34 @@ def reinitialize_distributed(
         if reconfig_request.new_data_parallel_rank == \
         ReconfigureRankType.SHUTDOWN_CURRENT_RANK:
             self.shutdown()
-        return
\ No newline at end of file
+        return
+
+    def nixl_pull_kvcache(
```
Comment (ruisearch42) [inline vllm/v1/executor/ray_distributed_executor.py]: ditto: not mention nixl

**Review by ruisearch42 — COMMENTED** [review]
Thanks for the PR!
First round review

**Comment by R2-Y** [comment]
@njhill Hi Nick, I tried your merge of https://github.com/vllm-project/vllm/pull/23770, and the scheduler is still stuck in the WAITING_FOR_REMOTE_KVS status check, although it fixes the issue of being unable to load remote KVCache when PP > 1. After all pp stages have received the kv cache, the last pp stage will return ModelRunnerOutput with the finish_recving list information, but it seems that the final Ray dag execution result is incorrect, so when getting the result from the Future, the kv_connector_output information in the ModelRunnerOutput of the last pp stage is None, which causes the scheduler never trigger _update_from_kv_xfer_finished to update the finish_recving_kv_req_ids list, then WAITING_FOR_REMOTE_KVS  always True?

**Comment by njhill** [comment]
@R2-Y is PP + NixlConnector now working properly with non-Ray (mp executor)? Just wanted to understand whether the remaining issue is Ray-specific.

**Comment by R2-Y** [comment]
> @R2-Y is PP + NixlConnector now working properly with non-Ray (mp executor)? Just wanted to understand whether the remaining issue is Ray-specific.

yep，mp executor work fine. It's ray's issue. My current PR can skip Ray DAG when try to pull kvcache.

File: vllm/v1/executor/ray_distributed_executor.py [hunk 13]
```diff
@@ -106,4 +176,36 @@ def reinitialize_distributed(
...
+        scheduler_output,
+    ) :
+        """Pull KVCache on the Ray workers.
+
+        Args:
+            scheduler_output: The scheduler output to execute.
+
+        """
+        output_refs = [
+            (worker.pull_kvcache_ray.remote(scheduler_output)
```
Comment (ruisearch42) [inline vllm/v1/executor/ray_distributed_executor.py:192]: I think this part is problematic, generally we should avoid mixing ray remote calls with Compiled Graph execution. Also why don't we just do `pull_kvcache` (which is basically`kv_connector_no_forward`) as part of the model execution?

**Review by ruisearch42 — COMMENTED** [review]
btw, I'm testing out a solution as well
